src/devices/sound/ym2151.cpp

```diff
@@ -1736,10 +1736,7 @@ void ym2151_device::write(offs_t offset, u8 data)
 	if (offset & 1)
 	{
 		if (!m_reset_active)
-		{
-			m_stream->update();
-			write_reg(m_lastreg, data);
-		}
+			synchronize(TIMER_WRITE, data | (m_lastreg << 8));
```

Comment:
Using `device_timer` is really inefficient, particularly if it’s frequent.  Going via `device_timer` takes this path:
* Delegate call to out-of-line thunk to `device_timer` (unconditional branch)
* Virtual dispatch for `device_timer` (unconditional branch)
* Switch on timer ID (slow conditional branch – not good for the branch predictor)

Whereas using a delegate timer just goes through one non-virtual or virtual call from the delegate.

Is it really necessary for all these new `synchronize` calls to take the performance hit of going through `device_timer` as well as introducing the potential for timer ID conflicts in derived classes?  What’s the actual benefit you get in exchange for making the code slower and more error-prone?  I realise that for allocated timers you get slightly better debug log messages and save state registration names if you use the `device_timer` form, but anonymous timers like this aren’t saved and aren’t around long enough for that to matter.

Reply:
device_timers are in general better for saving (i.e., anonymous device timers can be saved), so I tend to prefer them. Today it doesn't matter for 0-duration synchronization timers but it could if we ever wanted to support instant saving from the debugger. I could be swayed away from them if there's consensus that this is really so bad. The overhead you mention seems relatively small compared to the cost of the timer itself.

Reply:
Anonymous timers still can’t be saved because they don’t exist when the save state structure is set up during startup.  Making them go through `device_timer` doesn’t change that.

It’s definitely significant overhead for things that are timer-heavy like SNES.  I already more than 10% better `-bench` performance out of SNES with SuperFX just by reducing branch mispredictions in the scheduler’s timer code.  I didn’t change the SNES timers to use delegate form while I was at it because it hurts logging, but in testing it made a pretty big difference.

The DAC devices in particular could end up calling `synchronize` an awful lot with your changes.  It’s better to minimise overhead.